core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -388,14 +388,16 @@ class KafkaApis(val requestChannel: RequestChannel,
     val produceRequest = request.body[ProduceRequest]
     val numBytesAppended = request.header.toStruct.sizeOf + request.sizeOfBodyInBytes
 
-    if (produceRequest.isTransactional) {
-      if (!authorize(request.session, Write, Resource(TransactionalId, produceRequest.transactionalId, LITERAL))) {
+    if (produceRequest.hasTransactionalRecords) {
+      val isAuthorizedTransactional = produceRequest.transactionalId != null &&
```

Comment:
What's the intended behaviour if `transactionalId == null && produceRequest.hasTransactionalRecords`?

Reply:
Returning `TRANSACTIONAL_ID_AUTHORIZATION_FAILED` seemed like the best bet `INVALID_REQUEST` might be another option. Currently, without this check, the authorizer raises an NPE. We could push the null checks into the authorizor as well.

Reply:
Makes sense. I guess one can say that the `null` transaction id is never authorized.